File: xray_fluent/happ_crypt.py
"""Расшифровка закрытых ссылок Happ (`happ://crypt`, `crypt2..crypt5`, `crypt5.1`).

Happ умеет делиться подписками в «зашифрованном» виде, чтобы адрес подписки не
был виден в самой ссылке. Формат восстановлен сообществом (ключи Happ извлечены
из APK и опубликованы в проектах-дешифраторах под открытыми лицензиями). Схема:

* ``crypt``..``crypt4`` — тело в base64, расшифровывается RSA-PKCS1v15 одним из
  четырёх приватных ключей (по одному на каждый режим), блоками по размеру ключа.
* ``crypt5`` (классический) — гибрид: несколько строковых перестановок,
  RSA-PKCS1v15 (ключ выбирается по 8-символьному «маркеру») отдаёт ключ
  ChaCha20-Poly1305, которым расшифровывается тело.
* ``crypt5.1`` — тот же гибрид с изменённой раскладкой полей и расширенным
  набором ключей; нужный ключ подбирается перебором.

Результат расшифровки — обычный текст: либо URL подписки (``https://...``),
либо готовый список ссылок (``vless://...`` и т.п.). Дальше он уходит в тот же
пайплайн импорта подписок, что и любой другой ответ провайдера.

Так как схема опирается на приватные ключи, извлечённые из Happ, публично
доступен только их ограниченный набор. Для ссылок ``crypt5.1`` с ключом, которого
нет в наборе, функция бросает :class:`HappKeyUnavailableError` с понятным
сообщением.
"""

from __future__ import annotations

import base64
import re

from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.serialization import load_der_private_key

from .happ_crypt_keys import CRYPT5_KEYS_B64, PKCS1_KEYS_B64

HAPP_SCHEME = "happ://"

_CRYPT_PREFIXES: tuple[tuple[str, int], ...] = (
    ("crypt5/", 4),
    ("crypt4/", 3),
    ("crypt3/", 2),
    ("crypt2/", 1),
    ("crypt/", 0),
)


class HappDecryptError(Exception):
    """Не удалось расшифровать ссылку Happ (битый формат, неизвестная схема)."""


class HappKeyUnavailableError(HappDecryptError):
    """Нет приватного ключа под данную ссылку (например, формат ``crypt5.1``)."""


# ─────────────────────────────────────────────────────────────────────────────
#  Определение ссылок
# ─────────────────────────────────────────────────────────────────────────────


def is_happ_link(text: str) -> bool:
    return str(text or "").strip().lower().startswith(HAPP_SCHEME)


def _split_crypt_prefix(text: str) -> tuple[int, str] | None:
    """Возвращает ``(ordinal, payload)`` для happ://crypt*-ссылки, иначе ``None``.

    Схема (``happ://``) и токен режима (``cryptN/``) сопоставляются без учёта
    регистра; тело подписки сохраняется как есть.
    """
    body = str(text or "").strip()
    if body[:len(HAPP_SCHEME)].lower() != HAPP_SCHEME:
        return None
    path = body[len(HAPP_SCHEME):]
    for prefix, ordinal in _CRYPT_PREFIXES:
        if path[:len(prefix)].lower() == prefix:
            return ordinal, path[len(prefix):]
    return None


def is_happ_crypt_link(text: str) -> bool:
    return _split_crypt_prefix(text) is not None


# ─────────────────────────────────────────────────────────────────────────────
#  Строковые / байтовые помощники
# ─────────────────────────────────────────────────────────────────────────────


def _b64decode(text: str) -> bytes:
    """base64, терпимый к url-safe алфавиту и отсутствию паддинга."""
    compact = text.replace("-", "+").replace("_", "/")
    compact = "".join(compact.split())
    return base64.b64decode(compact + "=" * (-len(compact) % 4))


def _swap_pairs(text: str) -> str:
    """Меняет местами соседние символы попарно: ``ABCD`` → ``BADC``."""
    chars = list(text)
    for i in range(0, len(chars) - 1, 2):
        chars[i], chars[i + 1] = chars[i + 1], chars[i]
    return "".join(chars)


def _block_pair_swap(text: str) -> str:
    """Каждый полный 4-символьный блок ``ABCD`` → ``CDAB``; хвост — как есть."""
    full = len(text) - (len(text) % 4)
    out: list[str] = []
    for offset in range(0, full, 4):
        out.append(text[offset + 2:offset + 4])
        out.append(text[offset:offset + 2])
    out.append(text[full:])
    return "".join(out)


# ─────────────────────────────────────────────────────────────────────────────
#  RSA / ChaCha20-Poly1305
# ─────────────────────────────────────────────────────────────────────────────

_key_cache: dict[str, object] = {}


def _wrap_pkcs1_in_pkcs8(der: bytes) -> bytes:
    """Оборачивает PKCS#1 RSAPrivateKey (DER) в PKCS#8 PrivateKeyInfo."""
    def _der_len(size: int) -> bytes:
        if size < 0x80:
            return bytes([size])
        raw = size.to_bytes((size.bit_length() + 7) // 8, "big")
        return bytes([0x80 | len(raw)]) + raw

    rsa_algo_id = bytes.fromhex("300d06092a864886f70d0101010500")
    octet = b"\x04" + _der_len(len(der)) + der
    body = b"\x02\x01\x00" + rsa_algo_id + octet
    return b"\x30" + _der_len(len(body)) + body


def _load_private_key(b64: str):
    cached = _key_cache.get(b64)
    if cached is not None:
        return cached
    der = base64.b64decode(b64)
    try:
        key = load_der_private_key(der, password=None)
    except ValueError:
        # crypt..crypt4 хранятся как PKCS#1 — оборачиваем в PKCS#8.
        key = load_der_private_key(_wrap_pkcs1_in_pkcs8(der), password=None)
    _key_cache[b64] = key
    return key


def _rsa_decrypt(b64_key: str, ciphertext: bytes) -> bytes:
    return _load_private_key(b64_key).decrypt(ciphertext, PKCS1v15())


def _chacha_decrypt(key: bytes, nonce: bytes, ciphertext: bytes) -> bytes:
    return ChaCha20Poly1305(key).decrypt(nonce, ciphertext, None)


# ─────────────────────────────────────────────────────────────────────────────
#  crypt..crypt4
# ─────────────────────────────────────────────────────────────────────────────


def _decrypt_crypt1to4(ordinal: int, payload: str) -> str:
    b64_key = PKCS1_KEYS_B64[ordinal]
    key = _load_private_key(b64_key)
    key_size = (key.key_size + 7) // 8
    cipher = _b64decode(payload)
    if not cipher or len(cipher) % key_size != 0:
        raise HappDecryptError("crypt: длина шифртекста не кратна размеру RSA-блока")
    parts = [
        key.decrypt(cipher[i:i + key_size], PKCS1v15())
        for i in range(0, len(cipher), key_size)
    ]
    return b"".join(parts).decode("utf-8")


# ─────────────────────────────────────────────────────────────────────────────
#  crypt5 (классический)
# ─────────────────────────────────────────────────────────────────────────────


def _finish_crypt5(nonce: bytes, url_b64: str, enc_str: str, key_b64: str) -> str:
    """Общий финал crypt5/crypt5.1: RSA → ключ ChaCha → расшифровка → URL."""
    rsa_plain = _rsa_decrypt(key_b64, _b64decode(enc_str)).decode("latin-1")
    chacha_key = _b64decode(_swap_pairs(rsa_plain))
    if len(chacha_key) != 32:
        raise HappDecryptError("crypt5: неверная длина ключа ChaCha20")
    intermediate = _chacha_decrypt(chacha_key, nonce, _b64decode(url_b64)).decode("utf-8")
    return _b64decode(_swap_pairs(intermediate)).decode("utf-8")


def _try_decrypt_crypt5_legacy(payload: str) -> str | None:
    """Классический crypt5: маркер (первые+последние 4 символа) → RSA-ключ.

    Возвращает ``None``, если маркер неизвестен либо структура не подходит, —
    тогда вызывающий пробует формат crypt5.1.
    """
    shuffled = _block_pair_swap(payload)
    if len(shuffled) < 8:
        return None
    marker = shuffled[:4] + shuffled[-4:]
    key_b64 = CRYPT5_KEYS_B64.get(marker)
    if key_b64 is None:
        return None
    body = shuffled[4:-4]
    if len(body) < 13:
        return None
    match = re.match(r"^\d+", body[12:])
    if not match:
        return None
    segment_len = int(match.group(0))
    packed = body[12 + match.end():]
    if len(packed) < 1 + segment_len:
        return None
    url_b64 = packed[1:1 + segment_len]
    enc_str = packed[1 + segment_len:]
    try:
        nonce = body[:12].encode("ascii")
        return _finish_crypt5(nonce, url_b64, enc_str, key_b64)
    except Exception:
        return None


# ─────────────────────────────────────────────────────────────────────────────
#  crypt5.1
# ─────────────────────────────────────────────────────────────────────────────


def _c51_block_pair_swap(region: str, length: int) -> str:
    out: list[str] = []
    for j in range(1, length + 1):
        block = (j - 1) // 4
        pos = (j - 1) % 4
        index = 4 * block + ((pos + 2) % 4)
        if index < len(region):
            out.append(region[index])
    return "".join(out)


def _c51_extract_nonce(payload: str) -> str:
    n = payload[4:16]
    return (
        n[2] + n[3] + n[0] + n[1] + n[6] + n[7]
        + n[4] + n[5] + n[10] + n[11] + n[8] + n[9]
    )


def _c51_selector(payload: str) -> str:
    if len(payload) < 10:
        return ""
    return (
        payload[2:4] + payload[0:2] + payload[-6:-4] + payload[-2:]
    ).lower()


def _c51_make_cipher_b64(enc_str: str, split_on_inner_equals: bool) -> str:
    trailing_start = len(re.sub(r"=+$", "", enc_str))
    eq_idx = enc_str.find("=")
    cipher_b64 = (
        enc_str[eq_idx + 1:]
        if split_on_inner_equals and 0 <= eq_idx < trailing_start
        else enc_str
    )
    cleaned = re.sub(r"^=+", "", cipher_b64)
    cleaned = re.sub(r"=+$", "", cleaned)
    return cleaned + "=" * (-len(cleaned) % 4)


def _c51_candidates(payload: str) -> list[tuple[str, str, str, bool]]:
    """Возможные разбиения payload на (nonce, url_b64, enc_str, split)."""
    nonce_str = _c51_extract_nonce(payload)
    candidates: list[tuple[str, str, str, bool]] = []
    seen: set[tuple] = set()

    def push(url_b64: str, enc_str: str, split: bool) -> None:
        if not url_b64 or len(enc_str) < 684:
            return
        key = (
            len(url_b64), url_b64[:16], url_b64[-16:], enc_str[:16], enc_str[-16:]
        )
        if key in seen:
            return
        seen.add(key)
        candidates.append((nonce_str, url_b64, enc_str, split))

    try:
        n = int(payload[18:20])
    except ValueError:
        n = 0
    if n > 0 and len(payload) >= 20 + n + 684:
        url_region = payload[20:20 + n]
        enc_region = payload[20 + n:20 + n + 684]
        skip = ((n - 1) // 4) * 4 + 1
        url_b64 = payload[17] + _c51_block_pair_swap(url_region, n - 1)
        enc_str = url_region[skip] + _c51_block_pair_swap(enc_region, 683)
        push(url_b64, enc_str, True)

    for trailer_len in range(4, 9):
        url_len = len(payload) - 20 - 684 - trailer_len
        if url_len <= 0:
            continue
        url_region = payload[20:20 + url_len]
        enc_region = payload[20 + url_len:20 + url_len + 684]
        if len(enc_region) != 684:
            continue
        url_b64 = _c51_block_pair_swap(url_region, url_len)
        enc_str = _c51_block_pair_swap(enc_region, 684)
        push(url_b64, enc_str, False)
        if url_b64.endswith("="):
            push(url_b64[1:] + "=", enc_str, False)

    return candidates


def _candidate_keys(selector: str) -> list[str]:
    """Ключи-кандидаты: сначала по селектору/семейству, затем все остальные."""
    keys: list[str] = []
    seen: set[str] = set()

    def add(value: str | None) -> None:
        if value and value not in seen:
            seen.add(value)
            keys.append(value)

    add(CRYPT5_KEYS_B64.get(selector))
    family = selector[:4]
    if family:
        for marker, value in CRYPT5_KEYS_B64.items():
            if marker.startswith(family):
                add(value)
    for value in CRYPT5_KEYS_B64.values():
        add(value)
    return keys


def _decrypt_crypt51(payload: str) -> str:
    selector = _c51_selector(payload)
    keys = _candidate_keys(selector)
    candidates = _c51_candidates(payload)
    preferred_values = {
        value
        for marker, value in CRYPT5_KEYS_B64.items()
        if marker == selector or (selector[:4] and marker.startswith(selector[:4]))
    }
    preferred_keys = [key for key in keys if key in preferred_values]
    fallback_keys = [key for key in keys if key not in preferred_values]

    def decrypt_with_keys(key_candidates: list[str]) -> str | None:
        for nonce_str, url_b64, enc_str, split in candidates:
            try:
                nonce = nonce_str.encode("ascii")
                cipher_b64 = _c51_make_cipher_b64(enc_str, split)
            except Exception:
                continue
            for key_b64 in key_candidates:
                try:
                    rsa_plain = _rsa_decrypt(key_b64, _b64decode(cipher_b64)).decode("latin-1")
                except Exception:
                    continue
                for shaped in (_swap_pairs(rsa_plain), rsa_plain):
                    try:
                        chacha_key = _b64decode(shaped)
                        if len(chacha_key) != 32:
                            continue
                        intermediate = _chacha_decrypt(
                            chacha_key, nonce, _b64decode(url_b64)
                        ).decode("utf-8")
                        return _b64decode(_swap_pairs(intermediate)).decode("utf-8")
                    except Exception:
                        continue
        return None

    decrypted = decrypt_with_keys(preferred_keys)
    if decrypted is not None:
        return decrypted

    # Fallback to Node.js emulation decryptor
    from pathlib import Path
    import shutil
    
    node_bin = shutil.which("node")
    if node_bin:
        emu_dir = Path(__file__).parent / "happ_emulator"
        cli_js = emu_dir / "decrypt_cli.js"
        if cli_js.exists():
            from .subprocess_utils import run_text_pumped, CREATE_NO_WINDOW, decode_output
            try:
                link = f"happ://crypt5/{payload}"
                result = run_text_pumped(
                    [node_bin, str(cli_js), link],
                    timeout=20.0,
                    creationflags=CREATE_NO_WINDOW,
                )
                if result.returncode == 0:
                    decrypted = decode_output(result.stdout).strip()
                    if decrypted:
                        return decrypted
            except Exception:
                pass

    decrypted = decrypt_with_keys(fallback_keys)
    if decrypted is not None:
        return decrypted

    raise HappKeyUnavailableError(
        "не удалось расшифровать happ://crypt5-ссылку: подходящий приватный ключ "
        f"отсутствует в наборе (формат crypt5.1, маркер «{selector}»)"
    )


def _decrypt_crypt5(payload: str) -> str:
    legacy = _try_decrypt_crypt5_legacy(payload)
    if legacy is not None:
        return legacy
    return _decrypt_crypt51(payload)


# ─────────────────────────────────────────────────────────────────────────────
#  Публичный API
# ─────────────────────────────────────────────────────────────────────────────


def decrypt_happ_link(link: str) -> str:
    """Расшифровывает ``happ://crypt*`` и возвращает URL/текст подписки.

    Бросает :class:`HappDecryptError` (или :class:`HappKeyUnavailableError`),
    если ссылку расшифровать нельзя.
    """
    parsed = _split_crypt_prefix(link)
    if parsed is None:
        raise HappDecryptError("неподдерживаемая ссылка (ожидалась happ://crypt*)")
    ordinal, payload = parsed
    if not payload:
        raise HappDecryptError("пустое тело happ://crypt-ссылки")
    try:
        if ordinal == 4:
            return _decrypt_crypt5(payload)
        return _decrypt_crypt1to4(ordinal, payload)
    except HappDecryptError:
        raise
    except Exception as exc:  # noqa: BLE001 - единый понятный класс ошибки
        raise HappDecryptError(f"ошибка расшифровки happ-ссылки: {exc}") from exc

File: android/core/config/src/main/java/com/lumen/core/config/crypto/HappCrypt.kt
package com.lumen.core.config.crypto

import java.io.ByteArrayOutputStream
import java.math.BigInteger
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.interfaces.RSAPrivateKey
import java.security.spec.PKCS8EncodedKeySpec
import java.util.Base64
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

open class HappDecryptError(message: String, cause: Throwable? = null) : Exception(message, cause)

class HappKeyUnavailableError(message: String, cause: Throwable? = null) : HappDecryptError(message, cause)

object HappCrypt {

    const val HAPP_SCHEME = "happ://"

    private val LEGACY_DIGIT_PATTERN = Pattern.compile("^\\d+")
    private val REGEX_TRAILING_EQUALS = Regex("=+$")
    private val REGEX_LEADING_EQUALS = Regex("^=+")

    private val CRYPT_PREFIXES = listOf(
        Pair("crypt5/", 4),
        Pair("crypt4/", 3),
        Pair("crypt3/", 2),
        Pair("crypt2/", 1),
        Pair("crypt/", 0)
    )

    fun isHappLink(text: String?): Boolean {
        return text?.trim()?.lowercase()?.startsWith(HAPP_SCHEME) == true
    }

    fun isHappCryptLink(text: String?): Boolean {
        return splitCryptPrefix(text) != null
    }

    fun decryptHappLink(link: String): String {
        val parsed = splitCryptPrefix(link)
            ?: throw HappDecryptError("Unsupported scheme (expected happ://crypt*)")
        val (ordinal, payload) = parsed
        if (payload.isEmpty()) {
            throw HappDecryptError("Empty body of happ://crypt link")
        }
        return try {
            if (ordinal == 4) {
                decryptCrypt5(payload)
            } else {
                decryptCrypt1To4(ordinal, payload)
            }
        } catch (e: HappDecryptError) {
            throw e
        } catch (e: Exception) {
            throw HappDecryptError("Happ link decryption error: ${e.message}", e)
        }
    }

    private fun splitCryptPrefix(text: String?): Pair<Int, String>? {
        val body = text?.trim() ?: return null
        if (body.length < HAPP_SCHEME.length || !body.substring(0, HAPP_SCHEME.length).equals(HAPP_SCHEME, ignoreCase = true)) {
            return null
        }
        val path = body.substring(HAPP_SCHEME.length)
        for ((prefix, ordinal) in CRYPT_PREFIXES) {
            if (path.length >= prefix.length && path.substring(0, prefix.length).equals(prefix, ignoreCase = true)) {
                return Pair(ordinal, path.substring(prefix.length))
            }
        }
        return null
    }

    private fun b64decode(text: String): ByteArray {
        var compact = text.replace("-", "+").replace("_", "/")
        compact = compact.filterNot { it.isWhitespace() }
        val padLen = (4 - (compact.length % 4)) % 4
        val padded = compact + "=".repeat(padLen)
        return try {
            Base64.getDecoder().decode(padded)
        } catch (e: Exception) {
            Base64.getUrlDecoder().decode(padded)
        }
    }

    private fun swapPairs(text: String): String {
        val chars = text.toCharArray()
        var i = 0
        while (i < chars.size - 1) {
            val tmp = chars[i]
            chars[i] = chars[i + 1]
            chars[i + 1] = tmp
            i += 2
        }
        return String(chars)
    }

    private fun blockPairSwap(text: String): String {
        val full = text.length - (text.length % 4)
        val sb = StringBuilder()
        var offset = 0
        while (offset < full) {
            sb.append(text.substring(offset + 2, offset + 4))
            sb.append(text.substring(offset, offset + 2))
            offset += 4
        }
        sb.append(text.substring(full))
        return sb.toString()
    }

    private fun wrapPkcs1InPkcs8(der: ByteArray): ByteArray {
        fun derLen(size: Int): ByteArray {
            if (size < 0x80) return byteArrayOf(size.toByte())
            val raw = BigInteger.valueOf(size.toLong()).toByteArray()
            val trimmed = if (raw[0] == 0.toByte() && raw.size > 1) raw.copyOfRange(1, raw.size) else raw
            return byteArrayOf((0x80 or trimmed.size).toByte()) + trimmed
        }
        val hexAlgo = "300d06092a864886f70d0101010500"
        val rsaAlgoId = ByteArray(hexAlgo.length / 2) { i ->
            hexAlgo.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
        val octet = byteArrayOf(0x04) + derLen(der.size) + der
        val body = byteArrayOf(0x02, 0x01, 0x00) + rsaAlgoId + octet
        return byteArrayOf(0x30) + derLen(body.size) + body
    }

    private val keyCache = ConcurrentHashMap<String, PrivateKey>()

    private fun loadPrivateKey(b64Key: String): PrivateKey {
        keyCache[b64Key]?.let { return it }
        val der = b64decode(b64Key)
        val keyFactory = KeyFactory.getInstance("RSA")
        val key = try {
            keyFactory.generatePrivate(PKCS8EncodedKeySpec(der))
        } catch (e: Exception) {
            val pkcs8Der = wrapPkcs1InPkcs8(der)
            keyFactory.generatePrivate(PKCS8EncodedKeySpec(pkcs8Der))
        }
        keyCache[b64Key] = key
        return key
    }

    private fun rsaDecrypt(b64Key: String, ciphertext: ByteArray): ByteArray {
        val privateKey = loadPrivateKey(b64Key)
        val cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding")
        cipher.init(Cipher.DECRYPT_MODE, privateKey)
        return cipher.doFinal(ciphertext)
    }

    private fun chachaDecrypt(key: ByteArray, nonce: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("ChaCha20-Poly1305")
        val keySpec = SecretKeySpec(key, "ChaCha20")
        val paramSpec = IvParameterSpec(nonce)
        cipher.init(Cipher.DECRYPT_MODE, keySpec, paramSpec)
        return cipher.doFinal(ciphertext)
    }

    private fun decryptCrypt1To4(ordinal: Int, payload: String): String {
        val b64Key = HappCryptKeys.PKCS1_KEYS_B64.getOrNull(ordinal)
            ?: throw HappDecryptError("No RSA key for crypt mode $ordinal")
        val key = loadPrivateKey(b64Key)
        val keySize = (((key as RSAPrivateKey).modulus.bitLength() + 7) / 8)
        val cipherBytes = b64decode(payload)
        if (cipherBytes.isEmpty() || cipherBytes.size % keySize != 0) {
            throw HappDecryptError("crypt: ciphertext length is not a multiple of RSA block size")
        }
        val cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding")
        cipher.init(Cipher.DECRYPT_MODE, key)

        val out = ByteArrayOutputStream()
        var offset = 0
        while (offset < cipherBytes.size) {
            val part = cipher.doFinal(cipherBytes, offset, keySize)
            out.write(part)
            offset += keySize
        }
        return String(out.toByteArray(), Charsets.UTF_8)
    }

    private fun finishCrypt5(nonce: ByteArray, urlB64: String, encStr: String, keyB64: String): String {
        val rsaPlainBytes = rsaDecrypt(keyB64, b64decode(encStr))
        val rsaPlain = String(rsaPlainBytes, Charsets.ISO_8859_1)
        val chachaKey = b64decode(swapPairs(rsaPlain))
        if (chachaKey.size != 32) {
            throw HappDecryptError("crypt5: invalid ChaCha20 key length (${chachaKey.size})")
        }
        val intermediateBytes = chachaDecrypt(chachaKey, nonce, b64decode(urlB64))
        val intermediateStr = String(intermediateBytes, Charsets.UTF_8)
        return String(b64decode(swapPairs(intermediateStr)), Charsets.UTF_8)
    }

    private fun tryDecryptCrypt5Legacy(payload: String): String? {
        val shuffled = blockPairSwap(payload)
        if (shuffled.length < 8) return null
        val marker = shuffled.substring(0, 4) + shuffled.substring(shuffled.length - 4)
        val keyB64 = HappCryptKeys.CRYPT5_KEYS_B64[marker] ?: return null
        val body = shuffled.substring(4, shuffled.length - 4)
        if (body.length < 13) return null

        val tail = body.substring(12)
        val matcher = LEGACY_DIGIT_PATTERN.matcher(tail)
        if (!matcher.find()) return null

        val segmentLenStr = matcher.group()
        val segmentLen = segmentLenStr.toIntOrNull() ?: return null
        val packed = tail.substring(segmentLenStr.length)
        if (packed.length < 1 + segmentLen) return null

        val urlB64 = packed.substring(1, 1 + segmentLen)
        val encStr = packed.substring(1 + segmentLen)
        return try {
            val nonce = body.substring(0, 12).toByteArray(Charsets.US_ASCII)
            finishCrypt5(nonce, urlB64, encStr, keyB64)
        } catch (e: Exception) {
            null
        }
    }

    private fun c51BlockPairSwap(region: String, length: Int): String {
        val sb = StringBuilder()
        for (j in 1..length) {
            val block = (j - 1) / 4
            val pos = (j - 1) % 4
            val index = 4 * block + ((pos + 2) % 4)
            if (index < region.length) {
                sb.append(region[index])
            }
        }
        return sb.toString()
    }

    private fun c51ExtractNonce(payload: String): String {
        if (payload.length < 16) return ""
        val n = payload.substring(4, 16)
        return "" + n[2] + n[3] + n[0] + n[1] + n[6] + n[7] + n[4] + n[5] + n[10] + n[11] + n[8] + n[9]
    }

    private fun c51Selector(payload: String): String {
        if (payload.length < 10) return ""
        return (payload.substring(2, 4) + payload.substring(0, 2) +
                payload.substring(payload.length - 6, payload.length - 4) +
                payload.substring(payload.length - 2)).lowercase()
    }

    private fun c51MakeCipherB64(encStr: String, splitOnInnerEquals: Boolean): String {
        val trailingStart = encStr.replace(REGEX_TRAILING_EQUALS, "").length
        val eqIdx = encStr.indexOf('=')
        val cipherB64 = if (splitOnInnerEquals && eqIdx in 0 until trailingStart) {
            encStr.substring(eqIdx + 1)
        } else {
            encStr
        }
        val cleaned = cipherB64.replace(REGEX_LEADING_EQUALS, "").replace(REGEX_TRAILING_EQUALS, "")
        val padLen = (4 - (cleaned.length % 4)) % 4
        return cleaned + "=".repeat(padLen)
    }

    private data class C51Candidate(
        val nonceStr: String,
        val urlB64: String,
        val encStr: String,
        val split: Boolean
    )

    private fun c51Candidates(payload: String): List<C51Candidate> {
        val nonceStr = c51ExtractNonce(payload)
        val candidates = mutableListOf<C51Candidate>()
        val seen = mutableSetOf<String>()

        fun push(urlB64: String, encStr: String, split: Boolean) {
            if (urlB64.isEmpty() || encStr.length < 684) return
            val key = "${urlB64.length}:${urlB64.take(16)}:${urlB64.takeLast(16)}:${encStr.take(16)}:${encStr.takeLast(16)}"
            if (seen.contains(key)) return
            seen.add(key)
            candidates.add(C51Candidate(nonceStr, urlB64, encStr, split))
        }

        val n = if (payload.length >= 20) payload.substring(18, 20).toIntOrNull() ?: 0 else 0
        if (n > 0 && payload.length >= 20 + n + 684) {
            val urlRegion = payload.substring(20, 20 + n)
            val encRegion = payload.substring(20 + n, 20 + n + 684)
            val skip = ((n - 1) / 4) * 4 + 1
            val urlB64 = payload[17] + c51BlockPairSwap(urlRegion, n - 1)
            val encStr = urlRegion[skip] + c51BlockPairSwap(encRegion, 683)
            push(urlB64, encStr, true)
        }

        for (trailerLen in 4..8) {
            val urlLen = payload.length - 20 - 684 - trailerLen
            if (urlLen <= 0) continue
            val urlRegion = payload.substring(20, 20 + urlLen)
            val encRegion = payload.substring(20 + urlLen, 20 + urlLen + 684)
            if (encRegion.length != 684) continue
            val urlB64 = c51BlockPairSwap(urlRegion, urlLen)
            val encStr = c51BlockPairSwap(encRegion, 684)
            push(urlB64, encStr, false)
            if (urlB64.endsWith("=")) {
                push(urlB64.substring(1) + "=", encStr, false)
            }
        }
        return candidates
    }

    private fun candidateKeys(selector: String): List<String> {
        val keys = mutableListOf<String>()
        val seen = mutableSetOf<String>()

        fun add(value: String?) {
            if (!value.isNullOrEmpty() && seen.add(value)) {
                keys.add(value)
            }
        }

        add(HappCryptKeys.CRYPT5_KEYS_B64[selector])
        val family = selector.take(4)
        if (family.isNotEmpty()) {
            for ((marker, value) in HappCryptKeys.CRYPT5_KEYS_B64) {
                if (marker.startsWith(family)) {
                    add(value)
                }
            }
        }
        for (value in HappCryptKeys.CRYPT5_KEYS_B64.values) {
            add(value)
        }
        return keys
    }

    private fun decryptCrypt51(payload: String): String {
        val selector = c51Selector(payload)
        val keys = candidateKeys(selector)
        val candidates = c51Candidates(payload)
        val preferredValues = HappCryptKeys.CRYPT5_KEYS_B64.filter { (marker, _) ->
            marker == selector || (selector.length >= 4 && marker.startsWith(selector.substring(0, 4)))
        }.values.toSet()

        val preferredKeys = keys.filter { preferredValues.contains(it) }
        val fallbackKeys = keys.filter { !preferredValues.contains(it) }

        fun decryptWithKeys(keyCandidates: List<String>): String? {
            for (cand in candidates) {
                val nonce = try {
                    cand.nonceStr.toByteArray(Charsets.US_ASCII)
                } catch (e: Exception) {
                    continue
                }
                val cipherB64 = try {
                    c51MakeCipherB64(cand.encStr, cand.split)
                } catch (e: Exception) {
                    continue
                }

                for (keyB64 in keyCandidates) {
                    val rsaPlain = try {
                        String(rsaDecrypt(keyB64, b64decode(cipherB64)), Charsets.ISO_8859_1)
                    } catch (e: Exception) {
                        continue
                    }
                    for (shaped in listOf(swapPairs(rsaPlain), rsaPlain)) {
                        try {
                            val chachaKey = b64decode(shaped)
                            if (chachaKey.size != 32) continue
                            val intermediateBytes = chachaDecrypt(chachaKey, nonce, b64decode(cand.urlB64))
                            val intermediateStr = String(intermediateBytes, Charsets.UTF_8)
                            return String(b64decode(swapPairs(intermediateStr)), Charsets.UTF_8)
                        } catch (e: Exception) {
                            continue
                        }
                    }
                }
            }
            return null
        }

        val decrypted = decryptWithKeys(preferredKeys)
        if (decrypted != null) return decrypted

        val fallbackDecrypted = decryptWithKeys(fallbackKeys)
        if (fallbackDecrypted != null) return fallbackDecrypted

        throw HappKeyUnavailableError(
            "Failed to decrypt happ://crypt5 link: private key unavailable in key set (format crypt5.1, marker '$selector')"
        )
    }

    private fun decryptCrypt5(payload: String): String {
        val legacy = tryDecryptCrypt5Legacy(payload)
        if (legacy != null) return legacy
        return decryptCrypt51(payload)
    }
}
